### ISSUE TITLE
Support to download and use dapr cli in private network

### ISSUE THREAD

**Issue by newbe36524** [dapr/cli#640]
### Background

1. In mainland China, it is difficult to download the package file on github directly
2. We consider that some users in a private network need to initialize and run dapr cli. Therefore, in this case, github and docker hub cannot be connected

Therefore, we consider that cli can be designed to support pulling related resources from private asset hosts and docker Registry. To solve the limitations of the network.

### Proposal

1. For the files deployed on github, you can consider using scripts to pack these files into a specific folder structure. Users can redeploy these files to the http server on the private network.
2. It supports specifying a private docker registry when download the images with dapr cli.

### Code work

1. When cli send requests to download resources on github, it needs to support replace with custom asset host addresses or uri prefix to download
2. cli needs to support pull images from private docker registry
3. Scripts such as install.ps1 and install.sh need to support download cli from specified address or custom address prefix.

@andi-souwer and I will work for this feature

Please give me some suggestion if you got a better idea.

**Comment by wcs1only** [comment]
I like 2. and it should be pretty straightforward to do.

If we're going to pack up resources to deploy to a webserver on a private network, we might as well pack them up as a standalone installer, I think.

**Comment by artursouza** [comment]
I agree that this is an issue that needs to be addressed to unblock adoption in China. I also agree that supporting private docker registry should be a no brainer and just add support for that.

The only part that I am concerned is "For the files deployed on github, you can consider using scripts to pack these files into a specific folder structure. Users can redeploy these files to the http server on the private network." All the files are released on GitHub, so I wonder what is the advantage to packing then in a single file. We can provide instructions but it seems that it should already be possible to bake such http server after copying the files from GitHub.

Another point is the use of a CDN. We could allow changing the domain for download to reach a CDN that works in China.

**Comment by newbe36524** [comment]
@wcs1only @artursouza 
I think I may not have described it very well, so I'll add a little more content here. The problem we need to solve actually consists of two parts.

1. downloading dapr cli
2. using dapr cli to install and deploy other resources, such as charts and docker images

For the first part.

As @artursouza  said. If it's just to support acceleration in China, then it's very easy to use a CDN or forward proxy to solve the network problem.
However, I would prefer to expand the problem to consider the case of a deployment in a fully private network. In such a network, all internal servers would not have access to the Internet.
I didn't mean to package them into a file, but simply to package them into an agreed-upon folder structure that could be more easily picked up by install.ps1 and install.sh.
With this structure, we can download files from github to an private http server. Then we can use install.ps1 and install.sh to download them.

Back to the China zone, if we implement this private network deployment approach. Then at this point, just imagine the China zone as a large private network, then you can use the same approach to build a similar http server in China and implement acceleration.

For the second part.

This part refers to some of the dapr cli commands like dapr init and dapr init -k. These commands also need to use some of the docker hub and github resources.
If we assume that these resources are not accessible on some private networks, then the following considerations need to be made.

1. pull images from dockerhub to the docker registry on the private network, and then dapr cli needs to support passing parameters to pull the docker registry on the private network
2. For the resources on github, you can actually use a similar approach as in the first part, pack them into an agreed folder structure, and then allow dapr cli to accept specific parameters to change the download address.

**Comment by pkedy** [comment]
As a proof of concept I created this:
https://github.com/pkedy/standalone
https://github.com/pkedy/standalone/releases

It is an offline installer. It installs the CLI, daprd & dashboard, loads the Placement, Redis and Zipkin docker images locally, creates the default config and component configurations, and starts the containers, same as `dapr init` would do. Just without requiring a network connection after the initial download.

This program does not currently support the Kubernetes install (`dapr init -k`). To add that support, it would be a similar approach of loading the containers, tagging them, pushing them to an alternate container registry, and running Helm from packaged charts.

**Comment by newbe36524** [comment]
> 

This is really amazing!
It would be great if the documentation could be enriched and the publicity enhanced.
- Add a readme
- Add a note to docs.dapr.io

**Comment by newbe36524** [comment]
> As a proof of concept I created this: https://github.com/pkedy/standalone https://github.com/pkedy/standalone/releases
> 
> It is an offline installer. It installs the CLI, daprd & dashboard, loads the Placement, Redis and Zipkin docker images locally, creates the default config and component configurations, and starts the containers, same as `dapr init` would do. Just without requiring a network connection after the initial download.
> 
> This program does not currently support the Kubernetes install (`dapr init -k`). To add that support, it would be a similar approach of loading the containers, tagging them, pushing them to an alternate container registry, and running Helm from packaged charts.

This is really amazing!
It would be great if the documentation could be enriched and the publicity enhanced.
- Add a readme
- Add a note to docs.dapr.io

**Comment by pkedy** [comment]
@newbe36524 Thanks! Still a work in progress though. The Dapr team would need to own/support it before mentioning on docs.dapr.io.

**Comment by Gsonovb** [comment]
Add custom proxy server function first ,

I tried to update and found that I couldn't access (TLS handshake timeout)

**Comment by shaibarlev** [comment]
Please consider supporting this feature, since the are many scenarios where there is no external network support or blocking tcp communication.

**Comment by newbe36524** [comment]
I'm very happy to see this issue added to the 1.7 projects.

I'm sorry that I don't have time to work on it personally at the moment.
If @pkedy or someone else can continue this work. That would be great.

**Comment by artursouza** [comment]
I propose the scope this issue for the following work:

- [ ] Change CLI to be configurable to use a different registry and not only Docker Hub.
- [ ] Change GH Workflow to publish to GitHub Registry *in addition* to Docker Hub.
- [ ] Create a new workflow to clone (daily) 3rd party images from Docker Hub to GitHub Registry too: Zipkin, Redis, etc.
- [ ] Support Dapr init to work for standalone mode without pulling image from external registry, aka "airgap" environment.

Not in scope for this issue:
- Change test and quickstart images to use GH registry - these should be tracked via their own repositories.
- Support "airgap" install for K8s since K8s will always need an registry to pull images from, even if private.

**Comment by paulyuk** [comment]
> I propose the scope this issue for the following work:
> 
> * [ ]  Change CLI to be configurable to use a different registry and not only Docker Hub.
> * [ ]  Change GH Workflow to publish to GitHub Registry _in addition_ to Docker Hub.
> * [ ]  Create a new workflow to clone (daily) 3rd party images from Docker Hub to GitHub Registry too: Zipkin, Redis, etc.
> * [ ]  Support Dapr init to work for standalone mode without pulling image from external registry, aka "airgap" environment.
> 
> Not in scope for this issue:
> 
> * Change test and quickstart images to use GH registry - these should be tracked via their own repositories.
> * Support "airgap" install for K8s since K8s will always need an registry to pull images from, even if private.

I believe the standalone mode provides options for a wide range of private registries, which users and companies may need as a part of their policies, e.g. to have a certified secure supply chain.

**Comment by msfussell** [comment]
The proposal above makes sense. Comments
"Change CLI to be configurable to use a different registry and not only Docker Hub."
Please have a proposal on how this CLI will change. Presumably this is just for "dapr init" command

Although this is separate issue, it would be good to consider this making it easier to run the Sentry service locally at least doing this issue https://github.com/dapr/cli/issues/550 at the same time. Otherwise you are doing this local packaging for the Sentry service which is then not used (because it is hard to use locally)

**Comment by yaron2** [comment]
> I propose the scope this issue for the following work:
> 
> * [ ]  Change CLI to be configurable to use a different registry and not only Docker Hub.
> * [ ]  Change GH Workflow to publish to GitHub Registry _in addition_ to Docker Hub.
> * [ ]  Create a new workflow to clone (daily) 3rd party images from Docker Hub to GitHub Registry too: Zipkin, Redis, etc.
> * [ ]  Support Dapr init to work for standalone mode without pulling image from external registry, aka "airgap" environment.
> 
> Not in scope for this issue:
> 
> * Change test and quickstart images to use GH registry - these should be tracked via their own repositories.
> * Support "airgap" install for K8s since K8s will always need an registry to pull images from, even if private.

LGTM.

Until this is done, I suggest we put docs in place to explain how to install Dapr in air gapped environments today (ie. downloading the binary and putting it in your path as opposed to using `dapr init`, using in-memory state store/pubsub and/or the need to set up Redis and Zipkin from a local registry).

**Comment by mukundansundar** [comment]
LGTM.
